db/database.go

```diff
@@ -24,38 +25,116 @@ type Database interface {
 	Transact(fn TxFunc, opts ...TxOpt) (interface{}, error)
 	Metadata() (map[string]string, error)
 	Stats() sql.DBStats
+	WithContext(ctx context.Context) Database
+	WithLogger(logger logging.Logger) Database
+}
+
+type QueryOptions struct {
+	logger logging.Logger
+	ctx    context.Context
 }
 
 type SqlxDatabase struct {
-	db *sqlx.DB
+	db           *sqlx.DB
+	queryOptions *QueryOptions
 }
 
 func NewSqlxDatabase(db *sqlx.DB) *SqlxDatabase {
 	return &SqlxDatabase{db: db}
 }
 
+func (d *SqlxDatabase) getLogger() logging.Logger {
+	if d.queryOptions != nil {
+		return d.queryOptions.logger
+	}
+	return logging.Default()
+}
+
+func (d *SqlxDatabase) getContext() context.Context {
+	if d.queryOptions != nil {
+		return d.queryOptions.ctx
+	}
+	return context.Background()
+}
+
+func (d *SqlxDatabase) WithContext(ctx context.Context) Database {
+	return &SqlxDatabase{
+		db: d.db,
+		queryOptions: &QueryOptions{
+			logger: d.getLogger(),
+			ctx:    ctx,
+		},
+	}
+}
+
+func (d *SqlxDatabase) WithLogger(logger logging.Logger) Database {
+	return &SqlxDatabase{
+		db: d.db,
+		queryOptions: &QueryOptions{
+			logger: logger,
+			ctx:    d.getContext(),
+		},
+	}
+}
+
 func (d *SqlxDatabase) Close() error {
 	return d.db.Close()
 }
 
-func (d *SqlxDatabase) Get(dest interface{}, query string, args ...interface{}) error {
-	return d.db.Get(dest, query, args...)
+// reportFinish computes the duration since starts and logs a "done" report if that duration is
+// long enough.
+func (d *SqlxDatabase) reportFinish(err *error, fields logging.Fields, start time.Time) {
+	duration := time.Since(start)
+	if duration > 100*time.Millisecond {
+		d.getLogger().WithFields(fields).WithError(*err).WithField("duration", duration).Info("database done")
+	}
+}
+
+func (d *SqlxDatabase) Get(dest interface{}, query string, args ...interface{}) (err error) {
+	start := time.Now()
+	defer d.reportFinish(&err, logging.Fields{
```

Comment:
@itaiad200 also did not like this, but I'm unconvinced that the solution is to add more boilerplate to Go error handling.

I wrote it this way because I want to make error handling safe and explicit.  Try to write `Exec` without the `defer`-named-return-value trick.  I got this, which is very fragile:
```go
	start := time.Now()
	res, err := d.db.ExecContext(d.getContext(), query, args...)
	if err != nil {
		d.reportFinish(&err, logging.Fields{
				"type":  "exec",
				"query": query,
				"args":  args,
			}, start)
		return 0, err
	}

	rowsAffected, err := res.RowsAffected()
	d.reportFinish(&err, logging.Fields{
			"type":  "exec",
			"query": query,
			"args":  args,
		}, start)
	return rowsAffected
```
Now if you add any other early return you have to copy the `reportFinish` call.  This is repetition of the worst kind, the kind that humans forget about.
With TypeScript (or C++, maybe even with Java...!) I could pass a function-to-execute to `reportFinish`, with Python I could decorate. This way may require knowledge of Go to write the first time, but once written it can safely be modified.

Reply:
The idea is to capture the time and error in the highest level - anything that is more than one line can go though callback.
For example
```go
func (d *SqlxDatabase) Exec(query string, args ...interface{}) (count int64, err error) {
	res, err := d.performAndReport(logging.Fields{
		"type":  "exec",
		"query": query,
		"args":  args,
	}, func() (interface{}, error) {
		return d.db.ExecContext(d.getContext(), query, args...)
	})
	if err != nil {
		return 0, err
	}
	return res.(sql.Result).RowsAffected()
}

func (d *SqlxDatabase) performAndReport(fields logging.Fields, fn func() (interface{}, error)) (interface{}, error) {
	start := time.Now()
	data, err := fn()
	duration := time.Since(start)
	if duration <= 100*time.Millisecond {
		return data, err
	}
	l := d.getLogger().WithFields(fields).WithField("duration", duration)
	if err != nil {
		l = l.WithError(err)
	}
	l.Info("database done")
	return data, err
}
```

Reply:
Done :-(